### ISSUE TITLE
Wildcard performance

### ISSUE THREAD

**Issue by Volker-Weissmann** [mesonbuild/meson#10741]
From the FAQ:

> This means that a no-op build in a tree of 10 000 source files must take no more than a fraction of a second. This is only possible because Meson knows the exact list of files to check. If any target is specified as a wildcard glob, this is no longer possible. Meson would need to re-evaluate the glob every time and compare the list of files produced against the previous list. This means inspecting the entire source tree (because the glob pattern could be src/\*/\*/\*/\*.cpp or something like that). This is impossible to do efficiently.

Yes it can be done efficiently:
Find all directories that match `src/*/*/*/`, and put them at the end of this line:
```
build build.ninja: REGENERATE_BUILD ../meson.build ../meson_options.txt
```
If you create/remove a file inside the `src` tree, meson will reconfigure. 

Example performance measurement:
I cloned [llvm](https://github.com/llvm/llvm-project). 120k files in 12k directories. I added *all* directories, i.e.
```
find $PWD -type d | sed 's/ /$ /g' | tr '\n' ' '
```
as a dependency in an otherwise emtpy build.ninja file. This caused a no-op build to take 30 ms instead of 2 ms.

I confirmed that creating/deleting a file anywhere in the tree causes the target to be rebuild.

There is one disadvantage: If you use `some/path/*.c` and create `some/path/*.txt`, this would trigger a reconfigure.

**Comment by eli-schwartz** [comment]
> There is one disadvantage: If you use `some/path/*.c` and create `some/path/*.txt`, this would trigger a reconfigure.

And that's exactly the reason the FAQ says it cannot be done efficiently? No-op builds run a full reconfigure because something totally unrelated to the build, changed.

**Comment by Volker-Weissmann** [comment]
> > There is one disadvantage: If you use `some/path/*.c` and create `some/path/*.txt`, this would trigger a reconfigure.
> 
> And that's exactly the reason the FAQ says it cannot be done efficiently? No-op builds run a full reconfigure because something totally unrelated to the build, changed.

What if I think that's worth it? Can I get some method to add directories as a build.ninja dependency.

**Comment by xclaesse** [comment]
Also benchmarking should probably be done on Windows where the FS is many order of magnitude slower.

**Comment by Volker-Weissmann** [comment]
I wrote the patch below for my project, as a hacky workaround because no other options are available (see #10716)
Any chance of getting something like this merged into meson?
It's purely additive, you don't pay the price if you are not using it.

```
diff --git a/mesonbuild/backend/backends.py b/mesonbuild/backend/backends.py
index 63b873112..d9b79ef86 100644
--- a/mesonbuild/backend/backends.py
+++ b/mesonbuild/backend/backends.py
@@ -1186,6 +1186,7 @@ class Backend:
         definition needs to be regenerated.'''
         deps = OrderedSet([str(Path(self.build_to_src) / df)
                            for df in self.interpreter.get_build_def_files()])
+        deps.update(self.interpreter.regen_dirs)
         if self.environment.is_cross_build():
             deps.update(self.environment.coredata.cross_files)
         deps.update(self.environment.coredata.config_files)
diff --git a/mesonbuild/interpreter/interpreter.py b/mesonbuild/interpreter/interpreter.py
index 1dc80ad1d..9d9017350 100644
--- a/mesonbuild/interpreter/interpreter.py
+++ b/mesonbuild/interpreter/interpreter.py
@@ -101,6 +101,7 @@ import typing as T
 import textwrap
 import importlib
 import copy
+import glob
 
 if T.TYPE_CHECKING:
     import argparse
@@ -329,6 +330,7 @@ class Interpreter(InterpreterBase, HoldableObject):
         self.build_holder_map()
         self.user_defined_options = user_defined_options
         self.compilers: PerMachine[T.Dict[str, 'compilers.Compiler']] = PerMachine({}, {})
+        self.regen_dirs = set() # type: T.Set[str]
 
         # build_def_files needs to be defined before parse_project is called
         #
@@ -411,6 +413,7 @@ class Interpreter(InterpreterBase, HoldableObject):
                            'option': self.func_option,
                            'project': self.func_project,
                            'range': self.func_range,
+                           'regen_on_dir_change': self.func_regen_on_dir_change,
                            'run_command': self.func_run_command,
                            'run_target': self.func_run_target,
                            'set_variable': self.func_set_variable,
@@ -3275,3 +3278,25 @@ This will become a hard error in the future.''', location=self.current_node)
         if step < 1:
             raise InterpreterException('step must be >=1')
         return P_OBJ.RangeHolder(start, stop, step, subproject=self.subproject)
+
+    # todo: FeatureNew
+    # todo: Documentation
+    # todo: Unittest
+    @typed_kwargs('regen_on_dir_change', KwargInfo('recursive', (bool, NoneType), default=None))
+    @typed_pos_args('regen_on_dir_change', varargs=str)
+    def func_regen_on_dir_change(self, node: mparser.BaseNode, args: T.Tuple[str, bool],
+                                 kwargs: T.Dict[str, T.Any]) -> None:
+        mlog.warning('You are using regen_on_dir_change, which is a bit hacky and should only be used if there are no good other options.', once=True)
+        if kwargs['recursive'] is None:
+            raise InterpreterException('"recursive" kwarg missing')
+        recursive: bool = kwargs['recursive']
+        for dirp in args[0]:
+            src_root = self.environment.get_source_dir()
+            absbase_src = os.path.join(src_root, self.subdir)
+            absdir_src = os.path.join(absbase_src, dirp)
+            if not os.path.isdir(absdir_src):
+                raise InvalidArguments(f'Directory {dirp} does not exist.')
+            self.regen_dirs.add(absdir_src)
+            if recursive:
+                for el in glob.glob(f'{absdir_src}/*/', recursive=True):
+                    self.regen_dirs.add(el)
```
